src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicSession.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Security;
+using System.Text;
+
+namespace System.Net.Quic.Implementations.MsQuic.Internal
+{
+    internal sealed class MsQuicSession : IDisposable
+    {
+        private bool _disposed = false;
+        private IntPtr _nativeObjPtr;
+        private MsQuicApi _api;
+        private bool _opened;
+
+        internal MsQuicSession(string registration)
+        {
+            _api = new MsQuicApi();
+
+            _api.RegistrationOpen(Encoding.ASCII.GetBytes(registration));
```

Comment:
Why is the `MsQuicSession` opening a registration object? I would recommend against doing it this way.

Reply:
Sure, I'll refactor this to open the registration outside of the session.

Reply:
I'm having some trouble mapping Registrations and Sessions to this object model we'd like to have here. Ideally, we'd like consumers to not know anything about the registration and session when creating a connection or setting up a listener. For example:
```c#

QuicConnection connection = new QuicConnection(endpoint, sslClientOptions);
...
QuicListener listener = new QuicListener(endpoint, sslServerOptions);

```
Ideally these would internally setup both the QuicRegistration and QuicSession and "just work" as from a user standpoint, those should be transparent. However, as you mentioned, it's odd to be opening the registration from the session. Also as you mentioned:

> This way of dealing with sessions won't work correctly with 0-RTT/resumption. The app layer is going to need to take ownership of these most likely. Otherwise you need to keep a table of sessions, and if the app creates a new connection with the same ALPN, use the same previous session (unless they don't want to resume...?).

First off, the way I'm creating the sessions there is a hack 😄. Effectively, each connection/listener needs to have an associated session, which is where you pass in ALPN information. Session will be important for 0-RTT for connections, which is done by reusing the session. I'm not sure the importance in support 0-RTT at the moment, but I'd prefer to design around that potential constraint now.

For the Registration, I'm currently tempted to make everything static there or have a static singleton. @nibanks how important is it to call RegistrationClose after opening? IIRC you said it was fairly important, but if the process will be exiting at the point where we want to close the registration, is it still important to close (all sessions, listeners, connections and streams would be closed).

Reply:
cc @halter73 @geoffkizer

Reply:
Ok. First, for the registration, then you should just have a global API and Registration variable (lazily instantiated?). For session, you should have a global table of sessions that map to the `sslClientOptions`. If the client uses the same options (i.e. same ALPN) for a new connection, they should get the same session as before.

If you are existing, closing the registration isn't that important. If just your DLL/library is getting unloaded, then it's incredibly important. Can you tell the difference?

Reply:
> If you are exiting, closing the registration isn't that important. If just your DLL/library is getting unloaded, then it's incredibly important. Can you tell the difference?

Sounds good. I'll make this static for now then.

> BTW, I'd argue that 0-RTT is extremely important, as it's a primary feature of QUIC. You should design with it in mind right now.

Noted 😄

Reply:
> For session, you should have a global table of sessions that map to the sslClientOptions. If the client uses the same options (i.e. same ALPN) for a new connection, they should get the same session as before.

We'd also need to have a way to opt out of this behavior (ex maybe a bool on QuicConnection to reuse?)

Reply:
> > For session, you should have a global table of sessions that map to the sslClientOptions. If the client uses the same options (i.e. same ALPN) for a new connection, they should get the same session as before.
> 
> We'd also need to have a way to opt out of this behavior (ex maybe a bool on QuicConnection to reuse?)

sslClientOptions is a complex object with collections, callbacks, etc.. It does not make for a practical cache key.

Reply:
Then store the QUIC session in it.

Reply:
> BTW, I'd argue that 0-RTT is extremely important, as it's a primary feature of QUIC. You should design with it in mind right now.

Related, is there a way for us to control coalescing multiple smaller writes across different streams into maximally-sized packets?

Reply:
Not really. If you have several small buffers you want to queue, and want them in the same packet. Queue them with a single Send call.